android/expoview/src/main/java/host/exp/exponent/notifications/schedulers/CalendarSchedulerModel.java

```diff
@@ -0,0 +1,170 @@
+import com.cronutils.model.definition.CronDefinition;
+import com.cronutils.model.time.ExecutionTime;
+import com.cronutils.parser.CronParser;
+import com.raizlabs.android.dbflow.annotation.Column;
+import com.raizlabs.android.dbflow.annotation.PrimaryKey;
+import com.raizlabs.android.dbflow.annotation.Table;
+import com.raizlabs.android.dbflow.structure.BaseModel;
+
+import org.joda.time.DateTime;
+import org.json.JSONException;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+import host.exp.exponent.notifications.helpers.ExpoCronDefinitionBuilder;
+import host.exp.exponent.notifications.interfaces.SchedulerModelInterface;
+import host.exp.exponent.notifications.managers.SchedulersManagerProxy;
+import host.exp.exponent.notifications.managers.SchedulersDatabase;
+
+@Table(databaseName = SchedulersDatabase.NAME)
+public class CalendarSchedulerModel extends BaseModel implements SchedulerModelInterface {
+
+  private static List<String> triggeringActions = Arrays.asList(null,
+      Intent.ACTION_BOOT_COMPLETED,
+      Intent.ACTION_REBOOT,
+      Intent.ACTION_TIME_CHANGED,
+      Intent.ACTION_TIMEZONE_CHANGED);
+
+  private HashMap<String, Object> details;
+
+  // -- model fields --
+
+  @Column
+  @PrimaryKey(autoincrement = true)
+  int id;
+
+  @Column
+  int notificationId;
+
+  @Column
+  String experienceId;
+
+  @Column
+  boolean repeat;
+
+  @Column
+  String serializedDetails;
+
+  @Column
+  String calendarData;
+
+  // -- scheduler methods --
+
+  @Override
+  public String getIdAsString() {
+    return Integer.valueOf(id).toString() + this.getClass().getSimpleName();
+  }
+
+  @Override
+  public String getOwnerExperienceId() {
+    return experienceId;
+  }
+
+  @Override
+  public boolean canBeRescheduled() {
+    return repeat;
+  }
+
+  @Override
+  public String saveAndGetId() {
+    save(); // get id from database
+    details.put(SchedulersManagerProxy.SCHEDULER_ID, getIdAsString());
+    setDetails(details);
+    save();
+    return getIdAsString();
+  }
+
+  @Override
+  public void remove() {
+    delete();
+  }
+
+  public long getNextAppearanceTime() { // elapsedTime
+    CronDefinition cronDefinition = ExpoCronDefinitionBuilder.getCronDefinition();
+    CronParser parser = new CronParser(cronDefinition);
+    Cron cron = parser.parse(calendarData);
+
+    DateTime now = DateTime.now();
+    DateTime nextExecution = ExecutionTime.forCron(cron).nextExecution(now);
+    long whenShouldAppear = nextExecution.toDate().getTime();
+    long bootTime = DateTime.now().toDate().getTime() - SystemClock.elapsedRealtime();
+    return whenShouldAppear - bootTime;
+  }
+
+  @Override
+  public boolean shouldBeTriggeredByAction(String action) {
+    return CalendarSchedulerModel.triggeringActions.contains(action);
+  }
+
+  // model getters and setters
+
+  public int getId() {
+    return id;
+  }
+
+  public void setId(int id) {
+    this.id = id;
+  }
+
+  public int getNotificationId() {
```

Comment:
I think that since we do have nice interfaces in here, we might want to follow the rule: methods can be either `private`, `public static` or `@Override public`. Any other set of modifiers might indicate a code smell.